runtime/near-evm-runner/src/interpreter.rs

```diff
@@ -0,0 +1,275 @@
+
+use crate::evm_state::{EvmState, StateStore, SubState};
+use crate::near_ext::NearExt;
+use crate::types::Result;
+use crate::utils;
+
+pub fn deploy_code<T: EvmState>(
+    state: &mut T,
+    origin: &Address,
+    sender: &Address,
+    value: U256,
+    call_stack_depth: usize,
+    address_type: CreateContractAddress,
+    recreate: bool,
+    code: &[u8],
+) -> Result<Address> {
+    let mut nonce = U256::default();
+    if address_type == CreateContractAddress::FromSenderAndNonce {
+        nonce = state.next_nonce(&sender)?;
+    };
+    let (address, _) = utils::evm_contract_address(address_type, &sender, &nonce, &code);
+
+    if recreate {
+        state.recreate(address.0);
+    } else if state.code_at(&address)?.is_some() {
+        return Err(VMLogicError::EvmError(EvmError::DuplicateContract(hex::encode(address.0))));
+    }
+
+    let (result, state_updates) =
+        _create(state, origin, sender, value, call_stack_depth, &address, code)?;
+
+    // Apply known gas amount changes (all reverts are NeedsReturn)
+    // Apply NeedsReturn changes if apply_state
+    // Return the result unmodified
+    let (return_data, apply) = match result {
+        Some(GasLeft::Known(_)) => (ReturnData::empty(), true),
+        Some(GasLeft::NeedsReturn { gas_left: _, data, apply_state }) => (data, apply_state),
+        _ => return Err(VMLogicError::EvmError(EvmError::UnknownError)),
+    };
+
+    if apply {
+        state.commit_changes(&state_updates.unwrap())?;
+        state.set_code(&address, &return_data.to_vec())?;
+    } else {
+        return Err(VMLogicError::EvmError(EvmError::DeployFail(hex::encode(
+            return_data.to_vec(),
+        ))));
+    }
+    Ok(address)
+}
+
+pub fn _create<T: EvmState>(
+    state: &mut T,
+    origin: &Address,
+    sender: &Address,
+    value: U256,
+    call_stack_depth: usize,
+    address: &Address,
+    code: &[u8],
+) -> Result<(Option<GasLeft>, Option<StateStore>)> {
+    let mut store = StateStore::default();
+    let mut sub_state = SubState::new(sender, &mut store, state);
+
+    let params = ActionParams {
+        code_address: *address,
+        address: *address,
+        sender: *sender,
+        origin: *origin,
+        // TODO: gas usage.
+        gas: 1_000_000_000.into(),
+        gas_price: 1.into(),
+        value: ActionValue::Transfer(value),
+        code: Some(Arc::new(code.to_vec())),
+        code_hash: None,
+        data: None,
+        call_type: CallType::None,
+        params_type: vm::ParamsType::Embedded,
+    };
+
+    sub_state.transfer_balance(sender, address, value)?;
+
+    let mut ext = NearExt::new(*address, *origin, &mut sub_state, call_stack_depth + 1, false);
+    // TODO: gas usage.
+    ext.info.gas_limit = U256::from(1_000_000_000);
+    ext.schedule = Schedule::new_constantinople();
+
+    let instance = Factory::default().create(params, ext.schedule(), ext.depth());
+
+    // Run the code
+    let result = instance.exec(&mut ext);
```

Comment:
Do we count how many "evm" gas is used, calc and count it as part of near gas cost?

Reply:
No gas usage right now here at all. See TODOs that mention gas.

Reply:
I see, similar to i thought. For remap i'm thinking also measure it in runtime-param-estimator. Assume EVM it self says a1, a2, ... ak gas use, and runtime-param-estimator counting cpu instructions to be b1, b2, ... bk, a linear fit can tell how to transform this part. For precompile it's also measure by runtime-param-estimator. Other related thing i can think of:
- initialize evm, but when measure b1, ..., bk this should be part of c of y=mx+c
- post processing, apply state change, receipt, etc., this already count in your code